### PR TITLE
Feature: include category 

### PR DESCRIPTION
**Pull request by bnodir**
## Purpose

Adding "Include category" functionality in developer settings, with "All" as a default option. By combining data categorization with the `prepdocs.py` scripts, it lets you filter Azure AI Search based on categories. This functionality allows category-based inquiries and informs the user about the categories present in the search index.

This PR is somehow related to the following past issues:

- #2006 
- #1715
- #1333
- #872 
- #647 
- #493 

## Changes
- Modified approach.py and models.ts to include category logic
- Added translations for category ("All") in en, es, fr, and ja locales
- Updated Ask.tsx and Chat.tsx to handle category
- Updated data ingestion documentation

## Quick insight

_**Here is an example of how categories appear when ingested into the search index:**_

1. "category": "**_BenefitOptions_**"
2. "category": "**_RoleLibrary_**"
3. "category": "**_EmployeeHandbook_**"

image

_**Let's add category keys like this:**_
image

_**Then, in the default "All" category, we get answers from all three categories, like this:**_
image

_**When specifying a particular category, we get generated answers from that category's data:**_
image
image

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[ ] Yes
[x] No
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[x] Yes
[ ] No
```

## Type of change

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [ ] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

**Review by kalvarahul — APPROVED** [review]

**Comment by jeannot-prvlimburg** [comment]
Are there any updates on this feature?

File: infra/core/search/search-services.bicep [hunk 1]
```diff
@@ -70,4 +70,4 @@ resource search 'Microsoft.Search/searchServices@2023-11-01' = {
 output id string = search.id
 output endpoint string = 'https://${name}.search.windows.net/'
 output name string = search.name
-output principalId string = !empty(searchIdentityProvider) ? search.identity.principalId : ''
\ No newline at end of file
+output principalId string = !empty(searchIdentityProvider) ? search.identity.principalId : ''
```
Comment (pamelafox) [inline infra/core/search/search-services.bicep:73]: Were these EOL changes due to running `az bicep format` or from your editor?

File: docs/data_ingestion.md [hunk 2]
```diff
@@ -41,6 +42,12 @@ The script uses the following steps to index documents:
 3. Split the PDFs into chunks of text.
 4. Upload the chunks to Azure AI Search. If using vectors (the default), also compute the embeddings and upload those alongside the text.
 
+### Enhancing search functionality with data categorization
+
+To enhance search functionality, categorize data during the ingestion process with `--category` argument, for example `scripts/prepdocs.ps1 --category ExampleCategoryName`. This argument specifies the category to which the data belongs, enabling you to filter search results based on these categories.
+
+After running the script with the desired category, ensure these categories are added to the "Include Category" dropdown list in the developer settings. The default option for this dropdown is "All". By including specific categories, you can refine your search results more effectively.
```
Comment (pamelafox) [inline docs/data_ingestion.md:49]: Can you link to the relevant Chat/Ask files here, to make it clear what you mean by adding the categories?

File: app/frontend/src/pages/chat/Chat.tsx [hunk 3]
```diff
@@ -607,6 +615,30 @@ const Chat = () => {
                         )}
                     />
 
+                    <Dropdown
```
Comment (pamelafox) [inline app/frontend/src/pages/chat/Chat.tsx:618]: I wish that I had refactored these settings to be in a single component, as now we have duplicated code each time we add a new setting. This is fine for this PR, but eventually I want to try a single Settings.tsx across Chat/Ask.

File: app/frontend/src/locales/es/translation.json [hunk 4]
```diff
@@ -81,6 +81,10 @@
         "minimumSearchScore": "Puntaje mínimo de búsqueda",
         "minimumRerankerScore": "Puntaje mínimo de re-clasificación",
         "retrieveCount": "Obtén éste número resultados de búsqueda:",
+        "includeCategory": "Incluir categoría",
```
Comment (pamelafox) [inline app/frontend/src/locales/es/translation.json:84]: Let me know if any of these translations require a human review.

File: infra/core/search/search-services.bicep (hunk 1 above)
Comment (bnodir) [reply infra/core/search/search-services.bicep:73]: It's from my editor. After using the command `pre-commit run --all-files`, pre-commit hooks run on all the files in my repository.

File: app/frontend/src/pages/chat/Chat.tsx (hunk 3 above)
Comment (bnodir) [reply app/frontend/src/pages/chat/Chat.tsx:618]: Agreed, refactoring to a single file is a good future improvement (considered during the RAG HACK project). Let's proceed with this PR for now then.

File: app/frontend/src/locales/es/translation.json (hunk 4 above)
Comment (bnodir) [reply app/frontend/src/locales/es/translation.json:84]: They do need a human review, please.

File: docs/data_ingestion.md (hunk 2 above)
Comment (bnodir) [reply docs/data_ingestion.md]: Links added, please kindly re-review.

File: app/frontend/src/locales/es/translation.json [hunk 5]
```diff
@@ -128,6 +132,8 @@
             "Establece una puntuación mínima para los resultados de búsqueda que vuelven del re-clasificador semántico. La puntuación siempre varía entre 0-4. Cuanto mayor es la puntuación, más relevante es semánticamente el resultado a la pregunta.",
         "retrieveNumber":
             "Establece el número de resultados de búsqueda para recuperar de Azure AI search. Más resultados pueden aumentar la probabilidad de encontrar la respuesta correcta, pero pueden provocar que el modelo se 'pierda en el medio'.",
+        "includeCategory":
+            "Especifica una categoría para incluir en los resultados de búsqueda. No se utilizan categorías en el conjunto de datos predeterminado.",
```
Comment (glaucia86) [inline app/frontend/src/locales/es/translation.json:136]: LGTM

**Comment by pamelafox** [comment]
@glaucia86 Could you suggest the pt-br strings for "All" and "Include category" so @bnodir can add to the new pt-br file for this PR?

File: app/frontend/src/locales/es/translation.json (hunk 4 above)
Comment (glaucia86) [inline app/frontend/src/locales/es/translation.json:84]: LGTM

**Review by glaucia86 — APPROVED** [review]
For Spanish, sounds pretty good for me

**Review by pamelafox — CHANGES_REQUESTED** [review]
We need pt-br strings since I just merged that file, and I've asked humans to review the other strings.

**Comment by pamelafox** [comment]
@glaucia86 Thank you! Can you also provide the portuguese translations of those strings? The pt-br file isn't showing up in this PR since I just merged it in, but it needs updating with the same new strings.

File: app/frontend/src/locales/es/translation.json [hunk 6]
```diff
@@ -81,6 +81,10 @@
         "minimumSearchScore": "Puntaje mínimo de búsqueda",
         "minimumRerankerScore": "Puntaje mínimo de re-clasificación",
         "retrieveCount": "Obtén éste número resultados de búsqueda:",
```
Comment (alfredodeza) [inline app/frontend/src/locales/es/translation.json:83]:
```suggestion
        "retrieveCount": "Obtén éste número de resultados de la búsqueda:",
```

File: app/frontend/src/locales/es/translation.json [hunk 7]
```diff
@@ -81,6 +81,10 @@
...
+        "includeCategory": "Incluir categoría",
+        "includeCategoryOptions": {
+            "all": "Todos"
```
Comment (alfredodeza) [inline app/frontend/src/locales/es/translation.json:86]:
```suggestion
            "all": "Todas"
```

File: app/frontend/src/locales/es/translation.json (hunk 5 above)
Comment (alfredodeza) [inline app/frontend/src/locales/es/translation.json:136]:
```suggestion
            "Especifica una categoría para incluir en los resultados de la búsqueda. No se utilizan categorías en el conjunto de datos predeterminado.",
```

**Review by alfredodeza — CHANGES_REQUESTED** [review]

File: app/frontend/src/locales/fr/translation.json [hunk 8]
```diff
@@ -81,6 +81,10 @@
         "minimumSearchScore": "Score de recherche minimum",
         "minimumRerankerScore": "Score minimum du reclasseur sémantique",
         "retrieveCount": "Récupérer ce nombre de résultats de recherche :",
+        "includeCategory": "Inclure la catégorie",
+        "includeCategoryOptions": {
+            "all": "Tous"
```
Comment (manekinekko) [inline app/frontend/src/locales/fr/translation.json:86]: Is "all" referring to categories? If that the case, please accept the suggestion.
```suggestion
            "all": "Toutes"
```

File: app/frontend/src/locales/es/translation.json (hunk 7 above)
Comment (bnodir) [reply app/frontend/src/locales/es/translation.json:86]: https://github.com/Azure-Samples/azure-search-openai-demo/pull/2021#discussion_r1806787049 is accepted .

File: app/frontend/src/locales/es/translation.json (hunk 4 above)
Comment (bnodir) [reply app/frontend/src/locales/es/translation.json:84]: Changes made based on the review.

**Comment by bnodir** [comment]
Thank you, everyone. I have made the changes based on the reviewers' feedback.

**Comment by bnodir** [comment]
@zedhaque - Yes, the "All" option includes documents that don't have a category as well.

**Review by pamelafox — APPROVED** [review]

**Review by pamelafox — APPROVED** [review]
I think we *may* have some confusion with "All", with people wondering if that includes uncategorized, but we can watch and see if any issues come in about that. Thank you! I'll ask @mattgotteiner for a review as well since this adds a search filter.

**Comment by jeannot-prvlimburg** [comment]
I extended the list of choices in `app/frontend/src/locales/en/translation.json` to include:

`"includeCategoryOptions": {
    "all": "All",
    "digitalisering": "Digitalisering"
}`

However, only "All" appears as an option in the dropdown, even though I'm using English as the language. Could you advise on what might be missing for the dropdown to display the additional option(s)?

**Comment by cforce** [comment]
Does "All" also includes Uploaded Documents? It would be great to have as well "Uploaded" or "MyDocs"

**Comment by bnodir** [comment]
@jeannot-prvlimburg 

> I extended the list of choices in app/frontend/src/locales/en/translation.json to include:
> "includeCategoryOptions": { "all": "All", "digitalisering": "Digitalisering" }
> However, only "All" appears as an option in the dropdown, even though I'm using English as the language. Could you advise on what might be missing for the dropdown to display the additional option(s)?

Ensure your code looks like this:
```
                        aria-labelledby={includeCategoryId}
                        options={[
                            { key: "", text: t("labels.includeCategoryOptions.all") },   // ← Add comma
                            { key: "digitalisering", text: "Digitalisering" }   // ← Added new category line
                            // You can add a category key here for ingested data like below:
                            // { key: 'categoryName', text: 'Meaningful Category Name' }
                            // Alternatively, display the key to guide the user on what to type
                            // in the "Exclude category" field (e.g., 'Meaningful Category Name(categoryName)').
                        ]}
```

This should display the additional option in the dropdown. Additionally, please ensure to refresh your browser to see the changes.

**Comment by jeannot-prvlimburg** [comment]
> @jeannot-prvlimburg
> 
> > I extended the list of choices in app/frontend/src/locales/en/translation.json to include:
> > "includeCategoryOptions": { "all": "All", "digitalisering": "Digitalisering" }
> > However, only "All" appears as an option in the dropdown, even though I'm using English as the language. Could you advise on what might be missing for the dropdown to display the additional option(s)?
> 
> Ensure your code looks like this:
> 
> ```
>                         aria-labelledby={includeCategoryId}
>                         options={[
>                             { key: "", text: t("labels.includeCategoryOptions.all") },   // ← Add comma
>                             { key: "digitalisering", text: "Digitalisering" }   // ← Added new category line
>                             // You can add a category key here for ingested data like below:
>                             // { key: 'categoryName', text: 'Meaningful Category Name' }
>                             // Alternatively, display the key to guide the user on what to type
>                             // in the "Exclude category" field (e.g., 'Meaningful Category Name(categoryName)').
>                         ]}
> ```
> 
> This should display the additional option in the dropdown. Additionally, please ensure to refresh your browser to see the changes.

I found out indeed. Thanks for the help!

**Comment by jeannot-prvlimburg** [comment]
> @jeannot-prvlimburg
> 
> > I extended the list of choices in app/frontend/src/locales/en/translation.json to include:
> > "includeCategoryOptions": { "all": "All", "digitalisering": "Digitalisering" }
> > However, only "All" appears as an option in the dropdown, even though I'm using English as the language. Could you advise on what might be missing for the dropdown to display the additional option(s)?
> 
> Ensure your code looks like this:
> 
> ```
>                         aria-labelledby={includeCategoryId}
>                         options={[
>                             { key: "", text: t("labels.includeCategoryOptions.all") },   // ← Add comma
>                             { key: "digitalisering", text: "Digitalisering" }   // ← Added new category line
>                             // You can add a category key here for ingested data like below:
>                             // { key: 'categoryName', text: 'Meaningful Category Name' }
>                             // Alternatively, display the key to guide the user on what to type
>                             // in the "Exclude category" field (e.g., 'Meaningful Category Name(categoryName)').
>                         ]}
> ```
> 
> This should display the additional option in the dropdown. Additionally, please ensure to refresh your browser to see the changes.

Unfortunately, adding a line like this prevents the app from starting. Do you have any tips on resolving this?

**Comment by bnodir** [comment]
@jeannot-prvlimburg - I added the same code as above and didn't encounter any issues with the app starting. Could you elaborate a bit more on your situation? Specifically:

1. Did you edit the code, run `azd up`, and now your app is not starting at all? Can you share logs or screenshots?
2. Which release are you using? The latest release has significant changes, including switching from Azure App Services to Azure Container Apps.

### Recent Changes:
- [**2024-10-28**: Default to Container Apps instead of App Service](https://github.com/Azure-Samples/azure-search-openai-demo/releases/tag/2024-10-28)
- [**2024-10-2**3: New option to include category](https://github.com/Azure-Samples/azure-search-openai-demo/releases/tag/2024-10-23)

image

**Pull request merged**